tests/console/verify_efi_mok.pm

```diff
@@ -121,56 +112,49 @@ sub check_mok {
     diag('Expected regex used to verify SecureBoot: ' . $state);
     validate_script_output 'mokutil --sb-state', $state;
 
-    my $new_mok_keys = script_output('mokutil --list-new', proceed_on_failure => 1);
-    if ($new_mok_keys =~ /MokNew is empty/) {
+    if (script_output('mokutil --list-new', proceed_on_failure => 1) =~ /MokNew is empty/) {
         record_info 'MOK updates', 'No new certificates are expected to be enrolled';
-    } elsif ($new_mok_keys =~ /^Failed.*MokNew:.*directory/) {
-        record_soft_failure('bsc#1170889 -  mokutil: Failed to read MokNew');
     } else {
-        die "No new boot certificates are expected";
+        push @errors, 'No new boot certificates are expected';
     }
 
-    my $rc = script_run('mokutil --list-enrolled | grep ' . (is_sle() ? 'O=SUSE' : 'O=openSUSE'));
-    die "Unexpected certificate has been enrolled!\n" if ($rc && !get_var('DISABLE_SECUREBOOT', 0));
+    if (script_run(qq[mokutil --list-enrolled | tee /dev/$serialdev | grep -E "CN=.*SUSE"]) && !get_var('DISABLE_SECUREBOOT', 0)) {
+        push @errors, 'SUSE nor openSUSE certificate has not been found by mokutil';
+    }
 
-    unless (script_run 'test -f /boot/efi/EFI/mock.der') {
-        assert_script_run 'mokutil --list-enrolled | grep CN=MOCK';
-        assert_script_run 'mokutil --delete /boot/efi/EFI/mock.der --root-pw';
-        assert_script_run 'rm /boot/efi/EFI/mock.der';
-        assert_script_run 'mokutil --list-delete | grep CN=MOCK';
+    # In 3rd test object with SecureBoot, tests creates, imports and enrolles a mock certificate that should be removed
+    unless (script_run "test -f ${\MOCK_CRT}") {
+        record_info 'MOK delete', 'Removing MOCK certificate';
+        assert_script_run 'mokutil --list-enrolled | grep -E "CN=MOCK"';
+        if (script_output("mokutil --delete ${\MOCK_CRT} --root-pw") =~ /SKIP:\s+${\MOCK_CRT}\s+is\s+not\s+in\s+MokList/) {
+            push @errors, 'CN=MOCK certificate has not been found in MokList';
+        }
+        assert_script_run "rm ${\MOCK_CRT}";
+        assert_script_run 'mokutil --list-delete';
     }
 }
 
-sub virtualized_block_type {
-    my $backend        = join('_', grep { $_ } (get_required_var('BACKEND'), get_var('VIRSH_VMM_FAMILY')));
+sub get_esp_info {
+    my @parted_data    = split /\n/, script_output 'parted --list --machine --script';
     my $blk_dev_driver = {
         qemu         => 'virtblk',
         svirt_xen    => 'xvd',
         svirt_hyperv => 'scsi'
     };
-
-    return $blk_dev_driver->{$backend};
-}
-
-sub get_esp_from_parted {
-    my @parted_data = split /\n/, script_output 'parted --list --machine --script';
+    my $virtualized_block_type = sub {
+        return $blk_dev_driver->{${\(join '_', grep /$_/, get_required_var 'BACKEND', get_var 'VIRSH_VMM_FAMILY')}};
```

Comment:
Why do you need this interpolation here?
```suggestion
        return $blk_dev_driver->{join ('_', grep (/$_/, get_required_var 'BACKEND', get_var 'VIRSH_VMM_FAMILY'))};
```

Reply:
Are you sure about the change in the `grep()`.  If I understood it right you try to eleiminate `undef` from the list before.
So it is:
```bash
FOOF__BARclemix@linux-7k5u:~> perl -e 'print join("_", grep(/$_/, "FOOF", undef, "BAR")) . $/;'
FOOF__BAR
```
vs
```
clemix@linux-7k5u:~> perl -e 'print join("_", grep {$_} "FOOF", undef, "BAR") . $/;'
FOOF_BAR
```

Reply:
Actually I don't understand why you changed this function, can you help me?

Reply:
> Are you sure about the change in the `grep()`. If I understood it right you try to eleiminate `undef` from the list before.
> So it is:
> 
> ```shell
> FOOF__BARclemix@linux-7k5u:~> perl -e 'print join("_", grep(/$_/, "FOOF", undef, "BAR")) . $/;'
> FOOF__BAR
> ```
> 
> vs
> 
> ```shell
> clemix@linux-7k5u:~> perl -e 'print join("_", grep {$_} "FOOF", undef, "BAR") . $/;'
> FOOF_BAR
> ```

Actually `undef` yields warnings, so the `VIRSH_VMM_FAMILY` should return empty string, I will update that.
The thing is that each hypervisor has a different block device type. Therefore, I have this lookup table for it.
```
 my $blk_dev_driver = {
        qemu         => 'virtblk',
        svirt_xen    => 'xvd',
        svirt_hyperv => 'scsi'
    };
```my $a = 'qemu';
my $b = '';
my $blk_dev_driver = {
        qemu  => 'virtblk',
        svirt_xen => 'xvd',
        svirt_hyperv => 'scsi'
};
print "Testing \$a == $a and \$b == $b\n";
my $testfce = sub { return $blk_dev_driver->{join '_', grep /$_/, $a, $b}};
```

Got me
```
Testing $a == qemu and $b ==
T1: virtblk
```

Reply:
> ```shell
> perl -e 'print join("_", grep(/$_/, "FOOF", undef, "BAR")) . $/;'
> ```

Weird, I thought that `grep BLOCK LIST` or `grep EXPR,LIST` are equivalent forms. Let me do more tests. Thanks for the point!

Reply:
```
Another join test
Testing $a == qemu and $b ==
T1 EXPR: qemu
T1 BLOCK: qemu
Testing $a == qemu and $b ==  and $c == BAR
T1.3 EXPR: qemu_BAR
T1.3 BLOCK: qemu_BAR
```

```print "Another join test\n";
my $a = 'qemu';
my $b = '';
my $c = 'BAR';

print "Testing \$a == $a and \$b == $b\n";
my $testfce = sub { join '_', grep /$_/, $a, $b};
print "T1 EXPR: ${\$testfce->()}\n";
$testfce = sub { join '_', grep {$_} $a, $b};
print "T1 BLOCK: ${\$testfce->()}\n";
print "Testing \$a == $a and \$b == $b and \$c == $c\n";
$testfce = sub { join '_', grep /$_/, $a, $b, $c};
print "T1.3 EXPR: ${\$testfce->()}\n";
$testfce = sub { join '_', grep {$_} $a, $b, $c};
print "T1.3 BLOCK: ${\$testfce->()}\n";
```

Reply:
```
Testing $a == qemu and $b == undef and $c == BAR
Use of uninitialized value $_ in regexp compilation at grep_cap_test.pl line 69.
Use of uninitialized value $_ in pattern match (m//) at grep_cap_test.pl line 69.
T2.3 EXPR: qemu_BAR
T2.3 BLOCK: qemu_BAR
```

Reply:
Don't understand, I get different result. 
https://onlinegdb.com/rJoyI8buu

Reply:
The `for` loop is a good choice!